source/extensions/filters/http/cache/http_cache.cc

```diff
@@ -0,0 +1,125 @@
+#include "extensions/filters/http/cache/http_cache.h"
+
+#include <algorithm>
+
+#include "common/http/headers.h"
+#include "common/protobuf/utility.h"
+
+#include "extensions/filters/http/cache/http_cache_utils.h"
+
+#include "absl/time/time.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+LookupRequest::LookupRequest(const Http::HeaderMap& request_headers, SystemTime timestamp)
+    : timestamp_(timestamp),
+      request_cache_control_(request_headers.CacheControl() == nullptr
+                                 ? ""
+                                 : request_headers.CacheControl()->value().getStringView()) {
+  ASSERT(request_headers.Path(), "Can't form cache lookup key for malformed Http::HeaderMap "
+                                 "with null Path.");
+  ASSERT(request_headers.Scheme(), "Can't form cache lookup key for malformed Http::HeaderMap "
+                                   "with null Scheme.");
+  ASSERT(request_headers.Host(), "Can't form cache lookup key for malformed Http::HeaderMap "
+                                 "with null Host.");
+  const Http::HeaderString& scheme = request_headers.Scheme()->value();
+  const auto& scheme_values = Http::Headers::get().SchemeValues;
+  ASSERT(scheme == scheme_values.Http || scheme == scheme_values.Https);
+  // TODO(toddmgreer) Let config determine whether to include scheme, host, and
+  // query params.
+  // TODO(toddmgreer) get cluster name.
+  // TODO(toddmgreer) Parse Range header into request_range_spec_.
+  key_.set_cluster_name("cluster_name_goes_here");
+  key_.set_host(std::string(request_headers.Host()->value().getStringView()));
+  key_.set_path(std::string(request_headers.Path()->value().getStringView()));
+  key_.set_clear_http(scheme == scheme_values.Http);
+}
+
+// Unless this API is still alpha, calls to stableHashKey() must always return
+// the same result, or a way must be provided to deal with a complete cache
+// flush. localHashKey however, can be changed at will.
+size_t stableHashKey(const Key& key) { return MessageUtil::hash(key); }
+size_t localHashKey(const Key& key) { return stableHashKey(key); }
+
+// Returns true if response_headers is fresh.
+bool LookupRequest::isFresh(const Http::HeaderMap& response_headers) const {
+  const Http::HeaderEntry* cache_control_header = response_headers.CacheControl();
+  if (cache_control_header) {
+    const SystemTime::duration effective_max_age =
+        Internal::effectiveMaxAge(cache_control_header->value().getStringView());
+    return timestamp_ - Internal::httpTime(response_headers.Date()) < effective_max_age;
+  }
+  // We didn't find a cache-control header with enough info to determine
+  // freshness, so fall back to the expires header.
+  return timestamp_ <= Internal::httpTime(response_headers.get(Http::LowerCaseString("expires")));
```

Comment:
you can pull this out of Headers;:get() rather than constructing a new one, and it will then be an O(1) lookup.

Reply:
Done.